### PR TITLE
:sparkles: secret: Add generator for etcd client certs

### PR DESCRIPTION
**Pull request by randomvariable**
**What this PR does / why we need it**:
Adds a generator for etcd client certificates for use by the control plane controller to connect into the workload cluster's etcd endpoint.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #1902

### CODE REVIEW

File: util/secret/certificates.go [hunk 1]
```diff
@@ -215,17 +233,31 @@ func (c Certificates) Generate() error {
...
+			case ClusterAPIEtcdClient:
+				continue
 			default:
 				generator = generateCACert
 			}
-
-			kp, err := generator()
+			var kp *certs.KeyPair
+			var err error
+			if generator != nil {
```
Comment (chuckha) [inline util/secret/certificates.go]: `generator` cannot be `nil` given the switch statement. Why the change?

File: util/secret/certificates.go [hunk 2]
```diff
@@ -215,17 +233,31 @@ func (c Certificates) Generate() error {
...
+				kp, err = generator()
+			}
+			if certificate.Purpose == ClusterAPIEtcdClient {
```
Comment (chuckha) [inline util/secret/certificates.go]: i think this is impossible to reach? if the `certificate.Purpose == ClusterAPIEtcdClient` it would have already hit the `continue` block in the case statement.

Comment (randomvariable) [reply util/secret/certificates.go]: Messed that up. Have rewritten to match the intent.

File: util/secret/certificates.go [hunk 3]
```diff
@@ -395,6 +422,17 @@ func generateCACert() (*certs.KeyPair, error) {
 	}, nil
 }
 
+func generateClientCert(cn string, ca *certs.KeyPair) (*certs.KeyPair, error) {
```
Comment (chuckha) [inline util/secret/certificates.go]: these arguments are unused, either replace the names with `_` or remove the arguments entirely in order to clarify intent

File: util/secret/certificates.go [hunk 4]
```diff
@@ -410,6 +448,27 @@ func generateServiceAccountKeys() (*certs.KeyPair, error) {
 	}, nil
 }
 
+func newCertificate(cn string, ca *certs.KeyPair) (*x509.Certificate, *rsa.PrivateKey, error) {
```
Comment (chuckha) [inline util/secret/certificates.go]: this function is unused and unexported, is that intentional?

File: util/secret/certificates.go [hunk 5]
```diff
@@ -215,16 +233,25 @@ func (c Certificates) Generate() error {
 				continue
 			case ServiceAccount:
 				generator = generateServiceAccountKeys
+			case ClusterAPIEtcdClient:
+				if etcdCA == nil {
```
Comment (chuckha) [inline util/secret/certificates.go]: What do you think about adding a comment saying the EtcdCA certificate must come before ClusterAPIEtcdClient certificate in the Certificates type?

Comment (randomvariable) [reply util/secret/certificates.go]: Replaced with something less brittle.

**Comment by randomvariable** [comment]
/hold 
one final thing

File: util/secret/certificates.go [hunk 6]
```diff
@@ -215,10 +232,11 @@ func (c Certificates) Generate() error {
 				continue
 			case ServiceAccount:
 				generator = generateServiceAccountKeys
+			case ClusterAPIEtcdClient: // Depends on the etcd CA, so needs to be handled separately
```
Comment (dlipovetsky) [inline util/secret/certificates.go]: I found the `Generate` hard to understand; good candidate for a refactor in the future, I think.

File: util/secret/certificates.go [hunk 7]
```diff
@@ -456,3 +521,32 @@ func newSelfSignedCACert(key *rsa.PrivateKey) (*x509.Certificate, error) {
...
+
+	now := time.Now().UTC()
+
+	tmpl := x509.Certificate{
+		SerialNumber: new(big.Int).SetInt64(0),
+		Subject: pkix.Name{
+			CommonName:   cfg.CommonName,
+			Organization: cfg.Organization,
+		},
+		NotBefore:   now.Add(time.Minute * -5),
```
Comment (dlipovetsky) [inline util/secret/certificates.go]: Clever way to handle potential clock drift between management cluster and target host :smile:

File: util/secret/certificates.go [hunk 8]
```diff
@@ -456,3 +521,32 @@ func newSelfSignedCACert(key *rsa.PrivateKey) (*x509.Certificate, error) {
...
+		NotAfter:    now.Add(time.Hour * 24 * 365 * 10), // 10 years
```
Comment (dlipovetsky) [inline util/secret/certificates.go]: Worth defining as a constant somewhere?

Off-topic thought: I don't think we support user-defined expiration in general, but it might be a nice feature

File: util/secret/certificates.go [hunk 9]
```diff
@@ -456,3 +521,32 @@ func newSelfSignedCACert(key *rsa.PrivateKey) (*x509.Certificate, error) {
...
+
+func newClientCert(cn string, caCert *x509.Certificate, key *rsa.PrivateKey, caKey *rsa.PrivateKey) (*x509.Certificate, error) {
+	cfg := certs.Config{
+		CommonName: cn,
+	}
+
+	now := time.Now().UTC()
+
+	tmpl := x509.Certificate{
+		SerialNumber: new(big.Int).SetInt64(0),
```
Comment (dlipovetsky) [inline util/secret/certificates.go]: This may be pedantic...

>  The serial number MUST be a positive integer assigned by the CA to
>  each certificate.  It MUST be unique for each certificate issued by a
>  given CA (i.e., the issuer name and serial number identify a unique
>  certificate).  CAs MUST force the serialNumber to be a non-negative
>  integer.

-- https://tools.ietf.org/html/rfc3280#section-4.1.2.2

If we fix this, we could also fix it here: https://github.com/kubernetes-sigs/cluster-api/blob/f248367309a112b021c95bd8747b17200739c18a/util/secret/certificates.go#L437

File: util/secret/certificates.go (hunk 8 above)
Comment (dlipovetsky) [reply util/secret/certificates.go]: I see we define this in line here: https://github.com/kubernetes-sigs/cluster-api/blob/f248367309a112b021c95bd8747b17200739c18a/util/secret/certificates.go#L443

Comment (vincepri) [reply util/secret/certificates.go]: +1 for consts

File: util/secret/certificates.go (hunk 9 above)
Comment (randomvariable) [reply util/secret/certificates.go]: It looked wrong to me, but I didn't think to question it further.

File: util/secret/certificates.go (hunk 6 above)
Comment (randomvariable) [reply util/secret/certificates.go]: Yeah, it's painful/very organic.

**Comment by randomvariable** [comment]
The comments on this suggest it might as well be refactored now, and verified against the spec.

I'm going to close for the moment, and make sure this works with downstream consumption before reopening.

File: util/secret/certificates.go (hunk 8 above)
Comment (randomvariable) [reply util/secret/certificates.go]: +100
